be/src/udf/python/python_udf_server.py

```
@@ -860,14 +894,487 @@ def has_python_file_recursive(location: str) -> bool:
+            attr = inspect.getattr_static(udaf_class, "aggregate_state")
+            if not isinstance(attr, property):
+                raise ValueError(
+                    f"'aggregate_state' must be a @property in {udaf_class.__name__}"
+                )
+        except AttributeError:
+            raise ValueError(
+                f"UDAF class must have 'aggregate_state' property. "
+                f"Missing in {udaf_class.__name__}"
+            )
+
+
+class UDAFStateManager:
+    """
+    Manages UDAF aggregate states for Python UDAF execution.
+
+    This class maintains a mapping from place_id to UDAF instances,
+    following the Snowflake UDAF pattern:
+    - __init__(): Initialize state
+    - aggregate_state: Property returning serializable state
+    - accumulate(*args): Add input values
+    - merge(other_state): Merge two states
+    - finish(): Return final result
+    """
+
+    def __init__(self):
+        """Initialize the state manager."""
+        self.states: Dict[int, Any] = {}  # place_id -> UDAF instance
+        self.udaf_class = None  # UDAF class to instantiate
+        self.lock = threading.Lock()  # Thread-safe state access
+
+    def set_udaf_class(self, udaf_class: type):
+        """
+        Set the UDAF class to use for creating instances.
+
+        Args:
+            udaf_class: The UDAF class (must follow Snowflake pattern)
+        
+        Note:
+            Validation is performed by UDAFClassLoader before calling this method.
+        """
+        self.udaf_class = udaf_class
+        logging.info("UDAF class set to: %s", udaf_class.__name__)
+
+    def create_state(self, place_id: int) -> None:
+        """
+        Create a new UDAF state for the given place_id.
+
+        Args:
+            place_id: Unique identifier for this aggregate state
+        """
+        with self.lock:
+            if place_id in self.states:
+                # Destroy old state before creating new one
+                logging.warning(
```

should the logic happen ? maybe should return error?

---

yes, the logic should not happen. now raise instead of printing warning log